### PR TITLE
refactoring of mro_pm

### PR DESCRIPTION
**Pull request by gurneyalex**
- use more meaningful variables names in the planning strategy and the usage computation
- extract some time manipulation functions in a helpers submodule
- extract find_step to a function
- add a function field for meter lines to get the timestamp
- rewrite some loops using min() / max()
- replace multiple calls to dict.update with 1 call

### CODE REVIEW

**Comment by gurneyalex** [comment]
I did this refactoring while trying to better understand the logic of the code which was made difficult by the variable names. I'll be glad to work further on this if you think it's interesting (and to further change the identifiers if I got things wrong). 

The code is also difficult to test, mainly because of the implicit dates on meter lines. I did some manual testing but I may have missed some use cases.

**Comment by wowas18** [comment]
Thanks for your job. But I need time for review it. Please wait.

**Comment by gurneyalex** [comment]
no rush on my side. Thanks for considering this PR. 

Take your time, and if you only wish to include part of this that's fine with me too and I can help in rewriting the PR.

**Comment by Drsmartinkus** [comment]
Hi guys i notice a problem on  Assets & Finance By CodUP in combination with  Assets Management
By OpenERP S.A. module.  Assets Management By OpenERP S.A. module - module allows to validate suppliers invoice and at the same time to create a asset. But after installing Assets & Finance By CodUP it changes some of the fields in  Assets Management By OpenERP S.A. module. For example 
field "name" after installing Assets & Finance By CodUP module becames many2one relationship which is fine in terms of linking your assets with finance but when it comes to situation to validate suppliers invoice it raises the issue. Saying:
"null value in column "asset_id" violates not-null constraint
DETAIL: Failing row contains (3, 4, EXJCMMC/2015/000004, 2015-03-25 15:50:44.40533, null, f, null, 99, 150, number, 2015-03-25 15:50:44.40533, t, 2104, Volvo 45646, 99, 0.3, 100, 68, null, null, draft, 12, 2015-03-25, 26, linear, null).
" while evaluating
u'action_number()'

After further investigation i come to conclusion that Assets Management By OpenERP S.A. module tries to create a asset originally with simple char field "name" but now there is many2one relationship which caused the issue. So i assume it should work like this create asset.asset first and then link it with account.asset.asset

So in my point of view there some of the functionally within the system when developing this module wasn’t thought through.

**Comment by wowas18** [comment]
May be it is asset that was create before installing Assets & Finance. Try install on new DB without any demo data.

**Comment by Drsmartinkus** [comment]
Hi Thanks for the reply,

Yes your right i can bypass it now on clean database just wondering why on
other database it's still complaining about that new field added by Assets
& Finance asset_id is null ... Eventually i did a work around setting that
field to not required.

Thanks for help

File: mro_pm/mro.py [hunk 1]
```diff
@@ -37,128 +65,129 @@ class mro_order(osv.osv):
...
+        current_date = max(current_date, now)
+        current_meter_tot_val = meter_latest_tot_val + (current_date - latest_date_reading) * utilization_rate
+        delta = current_meter_tot_val - meter_daily_increase[-1]
         order_ids = self.search(cr, uid, 
             [('asset_id', '=', asset.id),
             ('state', '=', 'draft'),
             ('maintenance_type', '=', 'pm'),
             ('task_id', 'in', task_ids)],
             order='date_execution')
+        print "draft order_ids", order_ids
```
Comment (ljsalvatierra-binovo) [inline mro_pm/mro.py]: ¿print?

File: mro_pm/mro.py [hunk 2]
```diff
@@ -221,12 +261,12 @@ def planning_strategy_1(self, cr, uid, asset, meter, tasks, horizon, origin, con
...
-                if Dp > Dmax[i]: Dp = Dmax[i]
-            Co = Cp
-            Cp = C + (Dp - Dc)*N
-            delta = Cp - Co
+            current_date = date_opt[-1]
+            if nb_tasks > 1:
+                current_date = min(current_date, min(date_max[:-1]))
+            old_tot_val = current_meter_tot_val
+            current_meter_tot_val = meter_latest_tot_val + (current_date - latest_date_reading) * utilization_rate
+            delta = current_meter_tot_val - old_tot_val
```
Comment (ljsalvatierra-binovo) [inline mro_pm/mro.py]: Where is it used **_delta_**?

**Comment by ovnicraft** [comment]
Hi,  i am looking for implement this solution so this PR will me merged ?

**Pull request closed without merging**